butte-build/src/codegen.rs

```diff
@@ -0,0 +1,559 @@
+        let Self { path, doc } = self;
+        let stem = path
+            .file_stem()
+            .expect("invalid file stem")
+            .to_os_string()
+            .into_string()
+            .expect("invalid os string");
+        let id = format_ident!("{}", stem);
+        (quote! {
+            #doc
+            use #id;
+        })
+        .to_tokens(tokens)
+    }
+}
+
+// TODO:
+// * Root types
+// * File identifiers
+// * File extensions
+// * Parsing included files
+impl ToTokens for File<'_> {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        let Self { path, .. } = self;
+
+        // read the fbs file
+        let mut schema_text = String::new();
+        std::fs::File::open(path)
+            .expect("cannot open file")
+            .read_to_string(&mut schema_text)
+            .expect("cannot read file");
+
+        // parse the schema
+        let (_, Schema { includes, elements }) =
+            crate::parser::schema_decl(&schema_text[..]).expect("failed to parse");
+
+        // split the code into 0 or 1 namespace declarations and a body
+        let (namespaces, body): (Vec<_>, Vec<_>) = elements.into_iter().partition(|e| match e {
+            Element::Namespace(_) => true,
+            _ => false,
+        });
+
+        let (namespace_parts, doc): (Vec<_>, _) = match namespaces.len() {
+            0 => (vec![], Comment(None)),
+            1 => match &namespaces[0] {
+                Element::Namespace(Namespace { parts, doc }) => (parts.to_vec(), doc.clone()),
+                _ => unreachable!(),
+            },
+            _ => panic!("more than one namespace declaration found"),
+        };
+
+        // reverse fold over the namespace pieces to generate nested modules:
+        // if namespace parts is [a, b, c]
+        // and body is struct Foo { ... }
+        // then
+        // reverse:
+        //   [c, b, a]
+        // fold:
+        //    init: struct Foo { ... }
+        //    next: pub mod c { struct Foo { ... } }
+        //    next: pub mod b { pub mod c { struct Foo { ... } } }
+        //   final: pub mod a { pub mod b { pub mod c { struct Foo { ... } } } }
+        let schema_body =
+            namespace_parts
+                .iter()
+                .rev()
+                .fold(quote! { #(#body)* }, |module_body, module_name| {
+                    quote! {
+                        pub mod #module_name {
+                            #module_body
+                        }
+                    }
+                });
+
+        // finally, glom everything together
+        (quote! {
+            #(#includes)*
+            #doc // this is only filled if the namespace has a doc comment
+            #schema_body
+        })
+        .to_tokens(tokens)
+    }
+}
+
+#[test]
+fn test_visit_file() -> Result<(), Error> {
+    // TODO: not a huge fan of using env here, is there a better way?
+    let cargo_manifest_dir = env!("CARGO_MANIFEST_DIR");
```

Comment:
You should probably use `OUT_PATH`, see how bindgen does it: https://github.com/rust-lang/rust-bindgen/blob/master/book/src/tutorial-3.md

Reply:
This is for finding a `fbs` file to test, not for outputting generated code.

Reply:
I _am_ using `OUT_PATH` when generating code, FWIW.